ui/jobs/bars.ts

```diff
@@ -0,0 +1,784 @@
+import foodImage from '../../resources/ffxiv/status/food.png';
+import { UnreachableCode } from '../../resources/not_reached';
+import ResourceBar from '../../resources/resourcebar';
+import TimerBar from '../../resources/timerbar';
+import TimerBox from '../../resources/timerbox';
+import Util from '../../resources/util';
+import WidgetList, { Toward } from '../../resources/widget_list';
+import { Job } from '../../types/job';
+
+import { ShouldShows } from './components/base';
+import {
+  kMPCombatRate,
+  kMPNormalRate,
+  kMPTickInterval,
+  kMPUI1Rate,
+  kMPUI2Rate,
+  kMPUI3Rate,
+  kWellFedContentTypes,
+} from './constants';
+import { JobsEventEmitter } from './event_emitter';
+import './jobs_config';
+import { JobsOptions } from './jobs_options';
+import { Player } from './player';
+import { computeBackgroundColorFrom, makeAuraTimerIcon } from './utils';
+
+// text on the pull countdown.
+const kPullText = {
+  en: 'Pull',
+  de: 'Start',
+  fr: 'Attaque',
+  ja: 'タゲ取る',
+  cn: '开怪',
+  ko: '풀링',
+};
+
+type JobDomObjects = {
+  pullCountdown?: TimerBar;
+  leftBuffsList?: WidgetList;
+  rightBuffsList?: WidgetList;
+  cpBar?: ResourceBar;
+  gpBar?: ResourceBar;
+  healthBar?: ResourceBar;
+  manaBar?: ResourceBar;
+  mpTicker?: TimerBar;
+};
+
+export interface ResourceBox extends HTMLDivElement {
+  parentNode: HTMLElement;
+}
+
+export class Bars {
+  public o: JobDomObjects = {};
+
+  public ee: JobsEventEmitter;
+  public readonly player: Player;
+
+  constructor(private options: JobsOptions, o: {
+    emitter: JobsEventEmitter;
+    player: Player;
+  }) {
+    // Don't add any notifications if only the buff tracker is being shown.
+    if (this.options.JustBuffTracker) {
+      this.options.NotifyExpiredProcsInCombatSound = 'disabled';
+      this.options.NotifyExpiredProcsInCombat = 0;
+    }
+
+    this.ee = o.emitter;
+    this.player = o.player;
+
+    this.updateProcBoxNotifyRepeat();
+  }
+
+  updateProcBoxNotifyRepeat(): void {
+    if (this.options.NotifyExpiredProcsInCombat >= 0) {
+      const repeats = this.options.NotifyExpiredProcsInCombat === 0
+        ? 'infinite'
+        : this.options.NotifyExpiredProcsInCombat.toString();
+
+      document.documentElement.style.setProperty('--proc-box-notify-repeat', repeats);
+    }
+  }
+
+  _updateUIVisibility(hide?: boolean): void {
+    const bars = document.getElementById('bars');
+    if (bars)
+      bars.classList.toggle('pvp', hide ?? false);
+  }
+
+  _setupJobContainers(job: Job, shows: ShouldShows): void {
+    const shouldShows = {
+      buffList: true,
+      pullBar: true,
+      hpBar: true,
+      mpBar: true,
+      cpBar: false,
+      gpBar: false,
+      mpTicker: false,
+      ...shows,
```

Comment:
I'm not sure I understand this `ShouldShows` data flow.  Doesn't the call from `components/index.ts` set all of these values, so maybe `_setupJobContainers` should just take a `job` parameter and make the `Util` calls itself? Or, maybe just `const shouldShows = shows;` (or rename the parameter)?

Reply:
Hmm, my thought is that (as I wrote in the PR description) every job component has a `shouldShows` (or as you suggested above that should be `shouldShow`) and can override the default settings in `Bars._setupJobContainers` to control those widgets showing or not, but still, I have not completed it in a properly.
If the `shouldShow` field is designed as a non-static (I forgot what is it called in English sorry!) property, it is available only when the instance of the class created(constructed?), but for now "job components" initialise many "boxes" or "bars" in their constructors, which should be created after "hp-bar"/"mp-bar" or something else.
But if `shouldShow` field is designed as a static property, I don't know how to provide user customisation, it cannot be something like `BLMComponent.showShow = { /* blahblah */ }` which confused?